nav2_route/src/breadth_first_search.cpp

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2020 Samsung Research America
+// Copyright (c) 2023 Joshua Wallace
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_route/breadth_first_search.hpp"
+
+#include <iostream>
+#include <queue>
+#include <unordered_map>
+#include <vector>
+#include "nav2_core/planner_exceptions.hpp"
+#include "nav2_costmap_2d/cost_values.hpp"
+#include "nav2_costmap_2d/costmap_2d.hpp"
+#include "nav2_util/line_iterator.hpp"
+
+namespace nav2_route
+{
+
+void BreadthFirstSearch::initialize(nav2_costmap_2d::Costmap2D * costmap, int max_iterations)
+{
+  costmap_ = costmap;
+
+  int x_size = static_cast<int>(costmap_->getSizeInCellsX());
+  int y_size = static_cast<int>(costmap_->getSizeInCellsY());
+
+  max_index_ = x_size * y_size;
+  x_size_ = x_size;
+
+  neighbors_grid_offsets_ = {-1, +1,
+    -x_size, +x_size,
+    -x_size - 1, -x_size + 1,
+    +x_size - 1, +x_size + 1};
+
+  max_iterations_ = max_iterations;
+}
+
+BreadthFirstSearch::NodePtr BreadthFirstSearch::addToGraph(const unsigned int index)
+{
+  auto iter = graph_.find(index);
+  if (iter != graph_.end()) {
+    return &(iter->second);
+  }
+
+  return &(graph_.emplace(index, SimpleNode(index)).first->second);
+}
+
+void BreadthFirstSearch::setStart(unsigned int mx, unsigned int my)
+{
+  start_ = addToGraph(costmap_->getIndex(mx, my));
+}
+
+void BreadthFirstSearch::setGoals(std::vector<nav2_costmap_2d::MapLocation> & goals)
+{
+  goals_.clear();
+  for (const auto & goal : goals) {
+    goals_.push_back(addToGraph(costmap_->getIndex(goal.x, goal.y)));
+  }
+}
+
+void BreadthFirstSearch::search(unsigned int & goal)
+{
+  std::queue<NodePtr> queue;
+
+  start_->explored = true;
+  queue.push(start_);
+
+  int iteration = 0;
+  while (!queue.empty()) {
+    auto & current = queue.front();
+    queue.pop();
+
+    if (iteration > max_iterations_) {
+      throw nav2_core::PlannerTimedOut("Exceeded maximum iterations");
+    }
+
+
+    // Check goals
+    for (unsigned int i = 0; i < goals_.size(); ++i) {
```

Comment:
This is is what I mentioned about continuing until the cost is equal so that we get the closest one, not necessarily the first one since the search pattern isn't radial